opt: do not remap columns under a grouping operator

DrewKimball opened the pull request:
#### opt: do not remap columns under a grouping operator

The `TryRemapOuterCols` rules attempt to replace an outer column reference
within a relational expression by a non-outer column. The transformation is
valid because the rule only fires when a parent expression filters all rows
for which the outer and non-outer columns are not equal. For example:
```
SELECT * FROM xy INNER JOIN (SELECT a + x FROM ab) ON x = a;
```
In the above query, the reference to `x` in the subquery could be replaced
with `x`, and the rows for which `a = x` does not hold would be later
removed by the join.

However, the rule also performs this transformation for GroupBy and
DistinctOn, and in particular, the replacement column could be one of the
`ConstAgg` columns. This could cause incorrect results because input rows
that satisfy the equality and rows that don't could be aggregated together
(note that de-duplication is a special form of aggregation).

Note that the transformation is correct when the replacement column is one
of the grouping columns; this is because the grouping then separates rows
that satisfy the equality from those that don't. The groups that don't
satisfy the equality are then filtered later in the query, preserving the
correct result. This also applies to the distinct variants of set operators
(Union, Except, Intersect), since they de-duplicate across all columns.

This patch fixes the bug by restricting the rule to only make column
replacements under a GroupBy or DistinctOn for columns that are part of
the grouping column set.

Fixes #130001

Release note (bug fix): Fixed a bug introduced in v23.1 that can cause
incorrect results when:
1. The query contains a correlated subquery.
2. The correlated subquery has a GroupBy or DistinctOn operator with an
   outer-column reference in its input.
3. The correlated subquery is in the input of a Select or Join operator
4. The Select or Join has a filter that sets the outer-column reference from
   (2) equal to a non-outer column in the input of the grouping operator.
5. The grouping column set does not include the replacement column, and
   functionally determines the replacement column.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/130925)
<!-- Reviewable:end -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130925)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/opt/xform/testdata/external/tpch` line 86 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130925#-O73Jvv65XwZ46-t60VD:-O73Jvv65XwZ46-t60VE:b-u4fr61) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3719742786af5718b1569746f4a778148e6bc53/pkg/sql/opt/xform/testdata/external/tpch#L86)):*
> ```
> 
> # --------------------------------------------------
> # Q2
> ```

The plan changes here actually make the query faster; this is because the `GroupBy` is being pulled above a join due to decorrelation rules. It's a reversal of the regression mentioned here: https://github.com/cockroachdb/cockroach/pull/92378#issuecomment-1332742252


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/sql/opt/norm/decorrelate_funcs.go` line 1460 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130925#-O74807KETS0yTyEVDaZ:-O74807KETS0yTyEVDa_:b-28go2d) ([raw file](https://github.com/cockroachdb/cockroach/blob/f09d8777b3b1fbb2b9f0ed9e41ba3db36c6bad30/pkg/sql/opt/norm/decorrelate_funcs.go#L1460)):*
> ```Go
> 	case *memo.GroupByExpr, *memo.DistinctOnExpr:
> 		// [PushSelectIntoGroupBy]
> 		// Filters must refer only to grouping and ConstAgg columns.
> ```

Limiting this to just grouping columns, not ConstAgg columns, would also be a solution, right?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130925)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/opt/norm/decorrelate_funcs.go` line 1460 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130925#-O74807KETS0yTyEVDaZ:-O75KTYo-8Ys0kpwB7YR:b-y4vi4c) ([raw file](https://github.com/cockroachdb/cockroach/blob/f09d8777b3b1fbb2b9f0ed9e41ba3db36c6bad30/pkg/sql/opt/norm/decorrelate_funcs.go#L1460)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Limiting this to just grouping columns, not ConstAgg columns, would also be a solution, right?
</blockquote></details>

That's a good point, I think so. Actually, I think that might be justification for keeping the distinct set operators as well - their "grouping columns" are the set of all columns. I'll work on that.


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/sql/opt/norm/decorrelate_funcs.go` line 1460 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130925#-O74807KETS0yTyEVDaZ:-O766o_Z1vVfJTH91TLG:b-pdkh2o) ([raw file](https://github.com/cockroachdb/cockroach/blob/f09d8777b3b1fbb2b9f0ed9e41ba3db36c6bad30/pkg/sql/opt/norm/decorrelate_funcs.go#L1460)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

That's a good point, I think so. Actually, I think that might be justification for keeping the distinct set operators as well - their "grouping columns" are the set of all columns. I'll work on that.
</blockquote></details>

Cool. It makes sense why the `ConstAgg` columns are a problem: they may be consider constant because they are held equal to an outer column. Therefore, its invalid to replace equivalent outer columns with them because that removes the constraint that made them constant in the first place.


<!-- Sent from Reviewable.io -->

DrewKimball committed:
opt: do not remap columns under a grouping operator

The `TryRemapOuterCols` rules attempt to replace an outer column reference
within a relational expression by a non-outer column. The transformation is
valid because the rule only fires when a parent expression filters all rows
for which the outer and non-outer columns are not equal. For example:
```
SELECT * FROM xy INNER JOIN (SELECT a + x FROM ab) ON x = a;
```
In the above query, the reference to `x` in the subquery could be replaced
with `x`, and the rows for which `a = x` does not hold would be later
removed by the join.

However, the rule also performs this transformation for GroupBy and
DistinctOn, and in particular, the replacement column could be one of the
`ConstAgg` columns. This could cause incorrect results because input rows
that satisfy the equality and rows that don't could be aggregated together
(note that de-duplication is a special form of aggregation).

Note that the transformation is correct when the replacement column is one
of the grouping columns; this is because the grouping then separates rows
that satisfy the equality from those that don't. The groups that don't
satisfy the equality are then filtered later in the query, preserving the
correct result. This also applies to the distinct variants of set operators
(Union, Except, Intersect), since they de-duplicate across all columns.

This patch fixes the bug by restricting the rule to only make column
replacements under a GroupBy or DistinctOn for columns that are part of
the grouping column set.

Fixes #130001

Release note (bug fix): Fixed a bug introduced in v23.1 that can cause
incorrect results when:
1. The query contains a correlated subquery.
2. The correlated subquery has a GroupBy or DistinctOn operator with an
   outer-column reference in its input.
3. The correlated subquery is in the input of a Select or Join operator
4. The Select or Join has a filter that sets the outer-column reference from
   (2) equal to a non-outer column in the input of the grouping operator.
5. The grouping column set does not include the replacement column, and
   functionally determines the replacement column.

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130925)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/opt/norm/decorrelate_funcs.go` line 1460 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130925#-O74807KETS0yTyEVDaZ:-O76Dh8H8lxnsJ33CaF9:bz4kcqk) ([raw file](https://github.com/cockroachdb/cockroach/blob/f09d8777b3b1fbb2b9f0ed9e41ba3db36c6bad30/pkg/sql/opt/norm/decorrelate_funcs.go#L1460)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Cool. It makes sense why the `ConstAgg` columns are a problem: they may be consider constant because they are held equal to an outer column. Therefore, its invalid to replace equivalent outer columns with them because that removes the constraint that made them constant in the first place.
</blockquote></details>

Done. I think regressions will be minimal with this change.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Really great job working through this tricky one!

It might be nice to include a regression logic test that produced the incorrect result.

> Reviewed 2 of 8 files at r1, 6 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130925#-:-O76Z1m5D5x0eg_PsjZR:b-410ryy)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

DrewKimball commented:
>It might be nice to include a regression logic test that produced the incorrect result.

Done.

DrewKimball commented:
TFTR!

bors r+
